File: neetcode/347. Top K Frequent Elements/347.java
/*
Method 1.0: Using minHeap
Intution: build a frequency map and add key and count pairs to a minHeap, keep only the k most
frequent numbers.
TC: O(n log k)
SC: O(n + k)
 */
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();

        PriorityQueue<int[]> minHeap = new PriorityQueue<>(
                (a, b) -> a[1] == b[1] ? b[0] - a[0]: a[1] - b[1]);

        for (int i : nums) {
            map.put(i, map.getOrDefault(i, 0) + 1);
        }

        for (int key : map.keySet()) {
            minHeap.offer(new int[]{key, map.get(key)});
            if (minHeap.size() > k) {
                minHeap.poll();
            }
        }

        int[] result = new int[k];
        int index = 0;

        while (!minHeap.isEmpty()) {
            int[] temp = minHeap.poll();
            result[index++] = temp[0];
        }

        return result;
    }
}

File: neetcode/347. Top K Frequent Elements/347.kt
class Solution {
    fun topKFrequent(nums: IntArray, k: Int): IntArray {
        val map = HashMap<Int, Int>()

        val minHeap = PriorityQueue<Array<Int>>{ a, b -> if (a[1] == b[1]) b[0] - a[0] else a[1] - b[1]}

        for (num in nums) {
            map[num] = map.getOrDefault(num, 0) + 1
        }

        for (key in map.keys) {
            minHeap.offer(arrayOf(key, map[key]!!))
            if (minHeap.size > k) {
                minHeap.poll()
            }
        }

        val result = IntArray(k)
        var index = 0

        while (minHeap.isNotEmpty()) {
            result[index++] = minHeap.poll()[0]
        }

        return result
    }
}
